### PR TITLE
Add support for configured session variable to case claim 

### PR DESCRIPTION
**Pull request by orangejenny**
## Summary

This is a followup to https://github.com/dimagi/commcare-hq/pull/29004, but it's branched off of https://github.com/dimagi/commcare-hq/pull/29023/ because the latter PR does conflicting refactoring in `case_claim.js`

https://dimagi-dev.atlassian.net/browse/SUPPORT-7315

The new configurable session variable isn't compatible with the "Don’t claim cases already owned by user" checkbox, which uses a hard-coded expression that relies on `case_id`. This PR fixes that. It splits the `CaseSearch.relevant` attribute into a `default_relevant` and `additional_relevant` attribute, which matches the UI and gets rid of some ugly parsing code.

This includes a migration to migrate all `CaseSearch` objects. It took about 15 minutes to run on staging, and prod has 4 times as many documents. I'll talk to saas about whether or not this is too long a migration for a regular deploy and figure out an alternative if needed.

## Feature Flag
`Support several different case search/claim workflows in web apps: "search first", "see more", and "skip to default case search results"`

## Product Description
Nothing to announce beyond what's in https://github.com/dimagi/commcare-hq/pull/29023

## Safety Assurance

- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I am certain that this PR will not introduce a regression for the reasons below

### Automated test coverage

Present in this PR.

### QA Plan

QAed as part of https://github.com/dimagi/commcare-hq/pull/29023

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### CODE REVIEW

File: corehq/apps/app_manager/management/commands/migrate_case_search_relevant.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+class Command(AppMigrationCommandBase):
+    help = "One-time migration to replace CaseSearch.relevant with default_relevant and additional_relevant"
+
+    chunk_size = 20
+    include_builds = True
+    include_linked_apps = True
+
+    def migrate_app(self, app_doc):
+        default = "count(instance('casedb')/casedb/case[@case_id=instance('commcaresession')/session/data/case_id]) = 0"
```
[the remark this answers is not in the document]
Comment (orangejenny) [reply corehq/apps/app_manager/management/commands/migrate_case_search_relevant.py:16]: ignoring these

File: corehq/apps/app_manager/models.py [hunk 2]
```diff
@@ -2108,13 +2108,24 @@ class CaseSearch(DocumentSchema):
...
     include_closed = BooleanProperty(default=False)
     default_properties = SchemaListProperty(DefaultCaseSearchProperty)
     blacklisted_owner_ids_expression = StringProperty()
 
+    def get_relevant(self):
+        relevant = self.additional_relevant or ""
+        if self.default_relevant:
+            default_condition = CaseClaimXpath(self.session_var).default_relevant()
+            if relevant:
+                relevant = f"({relevant}) and ({default_condition})"
```
Comment (snopoke) [inline corehq/apps/app_manager/models.py:2124]: is the swapping of order here intentional? Previously the default was evaluated first (not sure if it makes any difference in the evaluation engine).

File: corehq/apps/app_manager/xpath.py [hunk 3]
```diff
@@ -320,6 +320,14 @@ def _gen(hierarchy):
         return [unicode_slug(t) for t in _gen(hierarchy)]
 
 
+class CaseClaimXpath(object):
+    def __init__(self, session_var_name):
+        self.session_var_name = session_var_name
+
+    def default_relevant(self):
+        return CaseIDXPath(session_var(self.session_var_name)).case().count().eq(0)
```
Comment (snopoke) [inline corehq/apps/app_manager/xpath.py:328]: :heart:

**Review by snopoke — APPROVED** [review]

File: corehq/apps/app_manager/models.py (hunk 2 above)
Comment (orangejenny) [reply corehq/apps/app_manager/models.py]: not intended, updated in 99b8c12c792955e89feef449bb7ed647f729907d

**Comment by orangejenny** [comment]
Switched base to master so this can be reverted if there are any problems with the migration without also reverting https://github.com/dimagi/commcare-hq/pull/29023

**Pull request merged**